### ISSUE TITLE
Error Output Results Possibly Related to partialRedundancyElimination(lastOptIndex=99)

### ISSUE THREAD

**Issue by Qeryu** [eclipse-openj9/openj9#19218]
Java -version output
--------------------

```bash
$ openj9-openjdk-jdk8/build/linux-x86_64-normal-server-release/images/j2sdk-image/bin/java -version
openjdk version "1.8.0_412-internal"
OpenJDK Runtime Environment (build 1.8.0_412-internal-b00)
Eclipse OpenJ9 VM (build master-a3f25b1fa, JRE 1.8.0 Linux amd64-64-Bit Compressed References 20240325_000000 (JIT enabled, AOT enabled)
OpenJ9   - a3f25b1fa
OMR      - cbfcf7d94
JCL      - 16aa7c2499 based on jdk8u412-b05)
```

Summary of problem
------------------

This should not be the same bug as [openj9-Issue #18777](https://github.com/eclipse-openj9/openj9/issues/18777). After pull [omr-Pull Request #7281](https://github.com/eclipse/omr/pull/7281), I encountered another issue with lastOptIndex being 99.

```java
class Test {
  long a;
  volatile double b;
  long c;
  void l(float d, int e, int f) {
    d = a;
    c += Float.floatToIntBits(d);
  }
  void g(int h) {
    int j = 3;
    int k = 2;
    l(1, j, 0);
    a += b = k;
  }
  void m(int n) {
    int o;
    for (o = 8; o < 348; o++)
      g(6);
  }
  void p(String[] q) {
    int i = 1;
    while (++i < 342)
      m(9);
    System.out.println(c);
  }
  public static void main(String[] r) {
    Test s = new Test();
    for (int i = 0; i < 10; i++)
      s.p(r);
  }
}
```

The issue can be reliably triggered using the command below.

```bash
$ java -Xjit:limit={Test.m*},lastOptIndex=99,optLevel=hot,optDetails,log=jit.log Test
138974768880640
276574232207872
276990892172832
277408854433952
277828006113984
278247876286464
278668375322944
279089503223424
279511259987904
279933509425440

# expected output
138974768880640
279890111208448
421553704461568
563660246227968
706171336112512
848926881248512
991896240144512
1135079412800512
1278476399216512
1422041040159104
```

Relevant jit.log information: 
[jit.log.2432521.15668.20240326.091428.2432521.log](https://github.com/eclipse-openj9/openj9/files/14751821/jit.log.2432521.15668.20240326.091428.2432521.log)

Additionally, I found that another test case mentioned in [openj9-Issue #18777](https://github.com/eclipse-openj9/openj9/issues/18777) can also continue to trigger the issue.

```java
class Test {
  int a = 400;
  volatile int b;
  int c;
  int[] iArrFld = new int[a];

  void d(String[] e) {
    int f = 6;
    do
      for (int h = 1; h < 100; ++h) {
        for (int j = 1; j < 2; j++)
          c -= 40;
        b = 0;
        iArrFld[1] <<= 2;
      }
    while (f++ < 200000);
    System.out.println(c);
  }

  public static void main(String[] k) {
    Test l = new Test();
    l.d(k);
  }
}
```

The run results are as follows. 

```bash
$ java -Xjit:limit={Test.d*},lastOptIndex=99,optLevel=hot,optDetails,log=jit.log Test
-117044920
# or -31770480
# or -36415880
# ...

# expected output
-791980200
```

Relevant jit.log information:
[jit.log.2430611.15580.20240326.091300.2430611.log](https://github.com/eclipse-openj9/openj9/files/14751823/jit.log.2430611.15580.20240326.091300.2430611.log)

**Comment by pshipton** [comment]
@hzongaro fyi

**Comment by hzongaro** [comment]
@a7ehuo, may I ask you to look at this one as well?  I know you've gained some familiarity with PRE after looking at #18777.  One thing I notice with this test is that it produces correct output if I remove the `volatile` attribute from `b`.

**Comment by a7ehuo** [comment]
I can reproduce this issue with `lastOptIndex=99,lastOptSubIndex=4` 

```
[   180]  99.4    O^O PARTIAL REDUNDANCY ELIMINATION: Eliminating redundant computation (lloadi) : 00007F4DF18878B0 in block_46, visit count 1156
```

The eliminated computation `0x7f4df18878b0` is `lloadi Test.c` which is part of the inlined code from `Test.l`. It is replaced by a `lload` from a temp initialized in loop pre-header.  I'm still looking.

```
CalleeIndex CallerIndex ByteCodeIndex CalleeMethod
       0         -1          13       Test.g(I)V
       1          0           8       Test.l(FII)V
       2          1          12       java/lang/Float.floatToIntBits(F)I
```
```
n414n     BBStart <block_46> (freq 10000) (in loop 46)                                        [0x7f4df188d120] bci=[2,0,744] rc=0 vc=1068 vn=109 li=-1 udi=- nc=0
n415n     astore  <temp slot 5>[#466  Auto] [flags 0x20000007 0x0 ] (X!=0 privatizedInlinerArg )  [0x7f4df188d170] bci=[-1,10,18] rc=0 vc=1068 vn=1 li=-1 udi=2 nc=1 flg=0x2004
n416n       aload  <'this' parm LTest;>[#421  Parm] [flags 0x40000107 0x0 ] (X!=0 X>=0 )      [0x7f4df188d1c0] bci=[-1,10,18] rc=1 vc=1068 vn=1 li=1 udi=28 nc=0 flg=0x104
n257n     astore  <temp slot 8>[#462  Auto] [flags 0x20000007 0x0 ] (X!=0 privatizedInlinerArg )  [0x7f4df188a010] bci=[0,4,12] rc=0 vc=1068 vn=1 li=-1 udi=3 nc=1 flg=0x2004
n82n        aload  <'this' parm LTest;>[#421  Parm] [flags 0x40000107 0x0 ] (X!=0 )           [0x7f4df1886960] bci=[0,4,12] rc=1 vc=1068 vn=1 li=1 udi=29 nc=0 flg=0x4
n129n     fstore  <temp slot 9>[#463  Auto] [flags 0x20000005 0x0 ]                           [0x7f4df1887810] bci=[1,5,6] rc=0 vc=1068 vn=13 li=-1 udi=4 nc=1
n128n       l2f                                                                               [0x7f4df18877c0] bci=[1,4,6] rc=1 vc=1068 vn=13 li=3 udi=- nc=1
n127n         lloadi  Test.a J[#443  Shadow +8] [flags 0x604 0x0 ] (cannotOverflow )          [0x7f4df1887770] bci=[1,1,6] rc=1 vc=1068 vn=12 li=2 udi=- nc=1 flg=0x1000
n126n           aload  <'this' parm LTest;>[#421  Parm] [flags 0x40000107 0x0 ] (X!=0 X>=0 )  [0x7f4df1887720] bci=[1,0,6] rc=1 vc=1068 vn=1 li=1 udi=30 nc=0 flg=0x104
n254n     lstore  <temp slot 7>[#461  Auto] [flags 0x4 0x0 ]                                  [0x7f4df1889f20] bci=[1,8,7] rc=0 vc=1068 vn=14 li=-1 udi=5 nc=1
n131n       lloadi  Test.c J[#446  Shadow +24] [flags 0x604 0x0 ] (cannotOverflow )           [0x7f4df18878b0] bci=[1,8,7] rc=1 vc=1068 vn=14 li=4 udi=- nc=1 flg=0x1000
n130n         aload  <'this' parm LTest;>[#421  Parm] [flags 0x40000107 0x0 ] (X!=0 X>=0 )    [0x7f4df1887860] bci=[1,6,7] rc=2 vc=1068 vn=1 li=1 udi=31 nc=0 flg=0x104
n253n     astore  <temp slot 6>[#460  Auto] [flags 0x7 0x0 ] (X!=0 )                          [0x7f4df1889ed0] bci=[1,6,7] rc=0 vc=1068 vn=1 li=-1 udi=6 nc=1 flg=0x4
n130n       ==>aload
n186n     istore  <auto slot 1>[#456  Auto] [flags 0x3 0x0 ]                                  [0x7f4df18889e0] bci=[2,4,744] rc=0 vc=1068 vn=15 li=-1 udi=7 nc=1
n180n       fbits2i                                                                           [0x7f4df1888800] bci=[2,1,744] rc=1 vc=1068 vn=15 li=6 udi=- nc=1
n179n         fload  <temp slot 9>[#463  Auto] [flags 0x20000005 0x0 ]                        [0x7f4df18887b0] bci=[2,0,744] rc=1 vc=1068 vn=13 li=5 udi=32 nc=0
n193n     ificmpeq --> block_21 BBStart at n191n ()                                           [0x7f4df1888c10] bci=[2,11,747] rc=0 vc=1068 vn=110 li=-1 udi=- nc=2 flg=0x20
n189n       iand (X>=0 cannotOverflow )                                                       [0x7f4df1888ad0] bci=[2,8,747] rc=1 vc=1068 vn=17 li=8 udi=- nc=2 flg=0x1100
n187n         iload  <auto slot 1>[#456  Auto] [flags 0x3 0x0 ] (cannotOverflow )             [0x7f4df1888a30] bci=[2,5,747] rc=1 vc=1068 vn=15 li=7 udi=33 nc=0 flg=0x1000
n188n         iconst 0x7f800000 (X!=0 X>=0 )                                                  [0x7f4df1888a80] bci=[2,6,747] rc=1 vc=1068 vn=16 li=-1 udi=- nc=0 flg=0x104
n190n       iconst 0x7f800000 (X!=0 X>=0 )                                                    [0x7f4df1888b20] bci=[2,9,747] rc=1 vc=1068 vn=16 li=-1 udi=- nc=0 flg=0x104
n178n     BBEnd </block_46> =====                                                             [0x7f4df1888760] bci=[2,11,747] rc=0 vc=920 vn=111 li=-1 udi=- nc=0
```

**Comment by a7ehuo** [comment]
Narrowed down the issue and testing a [potential fix](https://github.com/a7ehuo/omr/commit/d61ab55532a5d818d4750734f3e5b6d2e82b4ca7). Here are the details. 

The issue here is that PRE replaces the indirect load of field `Test.c` (`n131n`) in `block_46` with a temp `#471` because`li=4` is considered redundant. However, in the same loop, after `Test.c` is updated in `block_19` (`n143n`), there is no update on the temp `#471`. 

```
block_46:
  n131n lloadi Test.c #446 li=4 //<--- replaced by temp #471 by PRE
 
block_19:
  n143n lstorei Test.c #446 li=4 //<--- PRE doesn't create an update store to the temp #471 for #446 li=4

  n97n dstorei Test.b #444 li=15 volatile
```

The reason that the temp `#471` is not updated after `Test.c #446 li=4` is updated  in `block_19`  is because PRE doesn't think `li=4` is redundant computation in `block_19`. The redundant computation info for a block is [built based on `_downwardExposedAnalysisInfo`](https://github.com/eclipse/omr/blob/dd4c5d5459a1e2a890f18b68d5c6b7b105b658d0/compiler/optimizer/PartialRedundancy.cpp#L370).  

First, `li=4` is killed in `_transparencyInfo` for sym ref `#446`, which I think makes sense because the value for expression `li=4` changes if `#446` is changed. But `li=4` is not killed in `block_19` transparency info  because when LocalTransparency processes `#446`, [`li=4` store node survives](https://github.com/eclipse/omr/blob/57a22b20def101a9f997acc7eac45a7959fdb0a6/compiler/optimizer/LocalTransparency.cpp#L429-L462) even if it is killed in `_transparencyInfo[#446]`. 

Then when LocalAnticipatability processes the volatile store node `n97n` for `block_19` , `_downwardExposedAnalysisInfo` for `block_19` [intersects with](https://github.com/eclipse/omr/blob/57a22b20def101a9f997acc7eac45a7959fdb0a6/compiler/optimizer/LocalAnticipatability.cpp#L955-L965) all the killed sym refs transparency info including `_transparencyInfo[#446]`. So `li=4` is killed in `_downwardExposedAnalysisInfo` for `block_19`.

Now we have `li=4` can move across `block_19` based on local transparency, but it is not downward exposed. The inconsistency between the local transparency and local anticipability is the problem. Because `li=4` is not downward exposed in `block_19`, PRE doesn’t consider it redundant computation and it will not update the temp `#471` for `#446 li=4` in `block_19`.

```
Using transparency info for symRef #446

block_19 Merging with (_transparencyInfo[#446]): {0, 1, 2, 3, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27}
block_19 Defined (definedSymbolReferencesInBlock[block_19]): {23, 348, 376, 443, 446, 464}
block_19 Defined after stored (symRefsDefinedAfterStoredInBlock[block_19]): {}
block_19 Used after defined (symRefsUsedAfterDefinedInBlock[block_19]): {}
block_19 definedSymbolReferencesInBloc get(#446) 1 symRefsDefinedAfterStoredInBlock get(#446) 0 symRefsUsedAfterDefinedInBlock get(#446) 0
block_19 storeNode 4 supportedNodesAsArray[storeNode] n566n storeSymRefNum #446 child n567n childKilled 0 
block_19 symRefCanSurvive 1 (storeSymRefNum #446 == nextDefinedSymbolReference #446)
block_19 storeNode 4 supportedNodesAsArray n566n storeSymRefNum #446 nextDefinedSymbolReference #446 childKilled 0 symRefCanSurvive 1
block_19 Store node 4 n566n survives
block_19 _transparencyInfo[#446]): {0, 1, 2, 3, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27}
block_19 survivingStoreNodes: {4} //<---- li=4 survives
block_19 _analysisInfo: {1, 4, 5, 6, 7, 8, 9, 10, 11, 12, 14, 20, 21, 22, 23, 24, 25, 26, 27}
block_19 _analysisInfo: {1, 5, 6, 7, 8, 9, 10, 11, 12, 14, 20, 21, 22, 23, 24, 25, 26, 27}
block_19 _analysisInfo: {1, 4, 5, 6, 7, 8, 9, 10, 11, 12, 14, 20, 21, 22, 23, 24, 25, 26, 27}
```

```
TR_LocalTransparency: Solution for block number : block_19
block_19 _analysisInfo: {1, 4, 5, 6, 7, 8, 11, 14, 20, 21, 22, 23, 24, 25, 26, 27}
```

```
11Definition #15 (n97n) is locally anticipatable in block_19
updateUsesAndDefs: _registersScarce 0 node n97n node symRef #444 isAutoOrParm 0 symReference #444 sharesSymbol 1 seenStoredSymbolReferences 0
seenStoredSymbolReferences: {446, 458}
updateUsesAndDefs: calling killDownwardExposedExprs 3
killDownwardExposedExprs: block_19 node n97n li=15 (1)
tempContainer: {348, 376, 443, 444, 446, 464}
_downwardExposedAnalysisInfo: {1, 2, 4, 7, 9, 10, 11, 12, 13, 14, 15}
_downwardExposedStoreAnalysisInfo: {4, 15}
killDownwardExposedExprs: block_19 node n97n nodeBitShouldBeOn 1
killDownwardExposedExprs: block_19 node n97n storeNodeBitShouldBeOn 1
killDownwardExposedExprs: block_19 node n97n nextDefinedSymbolReference #348
tinfo: {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27}
_downwardExposedAnalysisInfo: {1, 2, 4, 7, 9, 10, 11, 12, 13, 14}
_downwardExposedStoreAnalysisInfo: {4}
killDownwardExposedExprs: block_19 node n97n nextDefinedSymbolReference #376
tinfo: {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27}
_downwardExposedAnalysisInfo: {1, 2, 4, 7, 9, 10, 11, 12, 13, 14}
_downwardExposedStoreAnalysisInfo: {4}
killDownwardExposedExprs: block_19 node n97n nextDefinedSymbolReference #443
tinfo: {0, 1, 4, 5, 6, 7, 8, 9, 10, 11, 12, 14, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27}
_downwardExposedAnalysisInfo: {1, 4, 7, 9, 10, 11, 12, 14}
_downwardExposedStoreAnalysisInfo: {4}
killDownwardExposedExprs: block_19 node n97n nextDefinedSymbolReference #444
tinfo: {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27}
_downwardExposedAnalysisInfo: {1, 4, 7, 9, 10, 11, 12, 14}
_downwardExposedStoreAnalysisInfo: {4}
killDownwardExposedExprs: block_19 node n97n nextDefinedSymbolReference #446
tinfo: {0, 1, 2, 3, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27}
_downwardExposedAnalysisInfo: {1, 7, 9, 10, 11, 12, 14} //<--- li=4 is killed 
_downwardExposedStoreAnalysisInfo: {}
killDownwardExposedExprs: block_19 node n97n nextDefinedSymbolReference #464
tinfo: {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27}
_downwardExposedAnalysisInfo: {1, 7, 9, 10, 11, 12, 14}
_downwardExposedStoreAnalysisInfo: {}
killDownwardExposedExprs: block_19 node n97n
_downwardExposedAnalysisInfo: {1, 7, 9, 10, 11, 12, 14, 15}
_downwardExposedStoreAnalysisInfo: {15}
```

```
TR_LocalAnticipatability: Solution for block number : block_19
block_19 _analysisInfo: {1, 2, 4, 7, 11, 13, 14, 15, 16, 17, 18, 19}
block_19 _downwardExposedAnalysisInfo: {1, 2, 7, 9, 10, 11, 12, 14}
block_19 _downwardExposedStoreAnalysisInfo: {2}
```
```
...
block_19 _rednSetInfo[19]: {1, 2, 14}
...
Redundancy info for block : 19 
{1, 2, 14}
```

@vijaysun-omr suggested a good idea of replacing the volatile store node with a call node and see what happens. In this experiment, I see `li=4` no longer survives `block_19` in local transparency. The key is that `symRefsDefinedAfterStoredInBlock` now contains the sym ref that `li=4` depends on. The volatile store should have been processed as [a call node in updateUsesAndDefs](https://github.com/eclipse/omr/blob/57a22b20def101a9f997acc7eac45a7959fdb0a6/compiler/optimizer/LocalTransparency.cpp#L592-L593) to have `symRefsDefinedAfterStoredInBlock` updated. 

```
n145n     lstorei  Test.c J[#448  Shadow +24] [flags 0x604 0x0 ]                              [0x7f2f0b0b1d10] bci=[1,17,8] rc=0 vc=1026 vn=19 li=4 udi=- nc=2
n259n       aload  <'this' parm LTest;>[#421  Parm] [flags 0x40000107 0x0 ] (X!=0 X>=0 )      [0x7f2f0b0b40b0] bci=[1,6,8] rc=1 vc=1026 vn=1 li=1 udi=35 nc=0 flg=0x104
n144n       ladd                                                                              [0x7f2f0b0b1cc0] bci=[1,16,8] rc=1 vc=1026 vn=19 li=12 udi=- nc=2
n143n         i2l                                                                             [0x7f2f0b0b1c70] bci=[1,15,8] rc=1 vc=1026 vn=18 li=10 udi=- nc=1
n255n           iload  <temp slot 5>[#460  Auto] [flags 0x3 0x0 ] (cannotOverflow )           [0x7f2f0b0b3f70] bci=[1,12,8] rc=1 vc=1026 vn=15 li=9 udi=36 nc=0 flg=0x1000
n260n         lload  <temp slot 7>[#463  Auto] [flags 0x4 0x0 ] (cannotOverflow )             [0x7f2f0b0b4100] bci=[1,8,8] rc=1 vc=1026 vn=14 li=11 udi=37 nc=0 flg=0x1000
n95n      treetop ()                                                                          [0x7f2f0b0b0d70] bci=[0,13,14] rc=0 vc=1026 vn=110 li=-1 udi=- nc=1 flg=0x8
n94n        calli  Test.x(I)V[#445  virtual Method -96] [flags 0x500 0x0 ] ()                 [0x7f2f0b0b0d20] bci=[0,13,14] rc=1 vc=1026 vn=109 li=- udi=- nc=3 flg=0x28
n93n          aloadi  <vft-symbol>[#345  Shadow] [flags 0x18607 0x0 ] (X!=0 )                 [0x7f2f0b0b0cd0] bci=[0,13,14] rc=1 vc=1026 vn=21 li=13 udi=- nc=1 flg=0x4
n91n            aload  <'this' parm LTest;>[#421  Parm] [flags 0x40000107 0x0 ] (X!=0 X>=0 )  [0x7f2f0b0b0c30] bci=[0,11,14] rc=2 vc=1026 vn=1 li=1 udi=38 nc=0 flg=0x104
n91n          ==>aload
n303n         iconst 6 (X!=0 X>=0 )                                                           [0x7f2f0b0b4e70] bci=[-1,11,21] rc=1 vc=1026 vn=20 li=-1 udi=- nc=0 flg=0x104
```
```
Using transparency info for symRef #448

block_19 Merging with (_transparencyInfo[#448]): {0, 1, 2, 3, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23}
block_19 Defined (definedSymbolReferencesInBlock[block_19]): {23, 348, 376, 446, 448, 466}
block_19 Defined after stored (symRefsDefinedAfterStoredInBlock[block_19]): {448}
block_19 Used after defined (symRefsUsedAfterDefinedInBlock[block_19]): {446}
block_19 definedSymbolReferencesInBloc get(#448) 1 symRefsDefinedAfterStoredInBlock get(#448) 1 symRefsUsedAfterDefinedInBlock get(#448) 0 //<--- li=4 depends on #448
block_19 _transparencyInfo[#448]): {0, 1, 2, 3, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23}
block_19 survivingStoreNodes: {} //<--- li=4 does not survive 
block_19 _analysisInfo: {1, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 17, 18, 19, 20, 21, 22, 23}
block_19 _analysisInfo: {1, 5, 6, 7, 8, 9, 10, 11, 12, 13, 17, 18, 19, 20, 21, 22, 23}
block_19 _analysisInfo: {1, 5, 6, 7, 8, 9, 10, 11, 12, 13, 17, 18, 19, 20, 21, 22, 23}
```

```
block_19 _analysisInfo: {1, 5, 6, 7, 8, 11, 13, 17, 18, 19, 20, 21, 22, 23}
```
